src/write/macho.rs

```diff
@@ -63,9 +63,112 @@ impl Object {
+                    value: 0,
+                    size: 0,
+                    kind: SymbolKind::Text,
+                    scope: SymbolScope::Dynamic,
+                    weak: false,
+                    section: None,
+                });
+                self.tlv_bootstrap = Some(id);
+                id
+            }
         }
     }
 
+    /// Create the `__thread_vars` entry for a TLS variable.
+    ///
+    /// The symbol given by `symbol_id` will be updated to point to this entry.
+    ///
+    /// A new `SymbolId` will be returned. The caller must update this symbol
+    /// to point to the initializer.
+    ///
+    /// If `symbol_id` is not for a TLS variable, then it is returned unchanged.
+    pub(crate) fn macho_add_thread_var(&mut self, symbol_id: SymbolId) -> SymbolId {
+        let symbol = self.symbol_mut(symbol_id);
+        if symbol.kind != SymbolKind::Tls {
+            return symbol_id;
+        }
+
+        // Create the initializer symbol.
+        let mut name = symbol.name.clone();
+        name.extend(b"$tlv$init");
+        let init_symbol_id = self.add_raw_symbol(Symbol {
+            name,
+            value: 0,
+            size: 0,
+            kind: SymbolKind::Tls,
+            scope: SymbolScope::Compilation,
+            weak: false,
+            section: None,
+        });
+
+        // Add the tlv entry.
+        // Three pointers in size:
+        //   - __tlv_bootstrap - used to make sure support exists
+        //   - spare pointer - used when mapped by the runtime
+        //   - pointer to symbol initializer
+        let section = self.section_id(StandardSection::TlsVariables);
+        let pointer_width = self.architecture.pointer_width().unwrap().bytes();
+        let size = u64::from(pointer_width) * 3;
+        let data = vec![0; size as usize];
+        let offset = self.append_section_data(section, &data, u64::from(pointer_width));
+
+        let tlv_bootstrap = self.macho_tlv_bootstrap();
+        self.add_relocation(
+            section,
+            Relocation {
+                offset: offset,
+                size: pointer_width * 8,
+                kind: RelocationKind::Absolute,
+                encoding: RelocationEncoding::Generic,
+                symbol: tlv_bootstrap,
+                addend: 0,
+            },
+        )
+        .unwrap();
+        self.add_relocation(
+            section,
+            Relocation {
+                offset: offset + u64::from(pointer_width) * 2,
+                size: pointer_width * 8,
+                kind: RelocationKind::Absolute,
+                encoding: RelocationEncoding::Generic,
+                symbol: init_symbol_id,
+                addend: 0,
+            },
+        )
+        .unwrap();
+
+        // Update the symbol to point to the tlv.
+        let symbol = self.symbol_mut(symbol_id);
+        symbol.value = offset;
+        symbol.size = size;
+        symbol.section = Some(section);
```

Comment:
Will this cause problems if you first use a relocation expecting it to point to the tls vars symbol, and then define the tls data, causing it to actually point to it?

Reply:
Not sure I understand.

If the user wants to add a relocation that refers to the tls symbol, then I think they should always use the symbol for `__thread_vars`, which will be the symbol they created and passed to `add_symbol_data`, so everything is ok there.

If the user wants to add a relocation inside the tls symbol data, then they need to add it to the `__thread_data` section, which is the section they passed to `add_symbol_data`, and they need to use the offset returned from `add_symbol_data`.

I think it'll only be a problem if the user reads `symbol.section` to determine which section to add the relocations to, instead of using the section they passed to `add_symbol_data`. I don't see a good way around that other than documenting it though.

Reply:
I reread the code together with your comment, and I get it. You can mark my other comment as resolved too.